### ISSUE TITLE
My rating isn't updating anymore

### ISSUE THREAD

**Issue by SvartfaR** [FAForever/server#546]
If you search for the 1v1 Leaderboards for "SvartfaR" it will show a rating of 411 points. This rating seems not to update since 4 further ranked games. Use the search of https://www.faforever.com/competitive/leaderboards/1v1 and take a look at the graph. The graph says it should be a rating of 705 points.

Maybe the ranked with replay id 11241961 caused the error. After this match I didn't got any notfication and the 1v1-Tab content didn't update as usually. (Thought my oppenent didn't leave the game.)

This issue occurs in the client and the website, so i think it is a server issue.
![Anmerkung 2020-03-19 170501](https://user-images.githubusercontent.com/51929253/77087864-eb19bc80-6a03-11ea-8e6b-faf7ca5b3116.png)
![Anmerkung 2020-03-19 170524](https://user-images.githubusercontent.com/51929253/77087866-ebb25300-6a03-11ea-91cf-4d978540a346.png)
![website](https://user-images.githubusercontent.com/51929253/77087868-ebb25300-6a03-11ea-9684-2a425f87e99d.png)

**Comment by Askaholic** [comment]
Hi, thanks for making the issue. I see that the results for both armies in the game you mentioned was `UNKNOWN` which explains why it wasn't rated. 

https://api.faforever.com/data/game?filter=(id==11241961)&include=playerStats

I need to get more recent logs before I can investigate why, but my guess is there was probably a conflict in the reported results. We've been trying to make the game rating more accurate, but it appears that in reality there are a rather large number of games that report inconsistent results and therefore get rejected (which would have been rated anyways by the old code).

If you could give a brief explanation of the outcome of the game here that would be useful. Include stuff like who won and how they won (was their ACU killed, did they ctrl-k, or did they leave the game?)

**Comment by SvartfaR** [comment]
Explanation of the outcome of the game: Horvat won because I ctrl-k my ACU and left the match via score screen.

> I see that the results for both armies in the game you mentioned was UNKNOWN which explains why it wasn't rated.

Do you mean it wasn't rated by the new code? In the graph you can see that I dropped from 461 to 412 points. (412 is next to 411, so I thought it's a rounding issue.)

Displays the graph the rating calculated by old code and the table displays the rating calculated of new code (which is stuck for me)?

**Comment by Askaholic** [comment]
Actually, I misread it. The `result` field isn't implemented yet so it will always show `UNKNOWN`. I see that there are 
```
afterDeviation | 149.63
afterMean | 860.394
```

which means that the game was rated. Also looking at your rating here:

https://api.faforever.com/data/player/16008?include=ladder1v1Rating

(copying it for documentation purposes)
```

data |  
-- | --
type | "player"
id | "16008"
attributes |  
createTime | "2012-03-12T20:00:38Z"
login | "SvartfaR"
updateTime | "2020-03-19T19:42:36Z"
userAgent | "downlords-faf-client"
relationships |  
avatarAssignments |  
data | []
bans |  
data | []
clanMembership |  
data | null
globalRating |  
data |  
type | "globalRating"
id | "16008"
ladder1v1Rating |  
data |  
type | "ladder1v1Rating"
id | "16008"
names |  
data | []
userNotes |  
data | []
included |  
0 |  
type | "ladder1v1Rating"
id | "16008"
attributes |  
deviation | 149.63
mean | 860.394
rating | 411.504
relationships |  
player |  
data |  
type | "player"
id | "16008"
```

we can see that it is still at these values. So I would say that game was rated correctly. 

Maybe you can supply the game id's of some of they games you played after this one, and we can check if those were rated as well.

**Comment by SvartfaR** [comment]
> Maybe you can supply the game id's of some of the games you played after this one, and we can check if those were rated as well.

In chronological order since the match 11241961:

1. 11242179: I won (probably) by killing enemy ACU (** 614)
2. 11242321: I lost, my ACU get killed (** 605)
3. 11242451: I won (probably) by killing enemy ACU (** 656)
4. 11242607: I won (probably) by killing enemy ACU (** 705)

** Rating after that match in the graph. All this ratings make sense, considering the enemies rating they were likely expected.
![Anmerkung 2020-03-19 213032](https://user-images.githubusercontent.com/51929253/77112200-e1ef1680-6a28-11ea-9e30-50c0d7b02fb7.png)

**Comment by Askaholic** [comment]
Hmm. Looks like they were all rated, but for some reason your rating in `ladder1v1_rating` wasn't updated.

I will note for myself here that these games were played while `v1.2` was still deployed so it is an issue that has been around for a while, not an issue introduced by our recent rating changes.

Have you played any more ladder games since then? Id be curious to see what happens if you play some more. I'm guessing you will see a weird jump in that rating graph.

**Comment by SvartfaR** [comment]
> Have you played any more ladder games since then?

No, I haven't.

Further information: Ingame and in the replay the score seemd to be correct:
![scoreboard](https://user-images.githubusercontent.com/51929253/77200938-e2041a80-6aeb-11ea-89b5-cfcbdb30ae18.png)

Does it have anything to do with it https://forums.faforever.com/viewtopic.php?f=12&t=18913 ?

**Comment by Askaholic** [comment]
Hah, I got it!  It's the same issue as #265. 

The score time for the first game was:
```
scoreTime | "2020-03-14T17:16:26Z"
```
which was later than the score times for all of your following games, the latest of those being:
```
scoreTime | "2020-03-14T17:02:53Z"
```

This also explains why you didn't get the notification asking you to leave a review after your game, since that notification only triggers after the game is rated.

Basically what must have happened was that your opponent in the first game either stayed on the score screen for a few hours or in some way lost connection or terminated the game without the `GameState: Ended` command reaching the server. Eventually when the connection timed out, or when he returned to his computer and closed the game, the game was finally rated, but because of the way the rating code works it used your old rating from when you started the game, rather than the new rating which you had acquired during that time.

Sadly there is nothing we can do to restore any rating points, you'll just have to play some more, but we are going to work on a fix for this and related issues in the near future. This was actually coming up on the todo list anyways.

**Comment by SvartfaR** [comment]
Today I played 1v1 ranked again. I started with a rating of 411, as expected. Now it's seems to work correct again.

I will double check now if the match counted before starting another one. :)

Thanks for your help Askaholic.

**Comment by Askaholic** [comment]
You're welcome! 

I think if you wait until the client gives you that notification then you will know that the game has been rated. I also recently merged #511 which isn't a direct fix but will help a lot with this problem once it gets deployed.
